### PR TITLE
Add Base32 & Base64 on encodeString

### LINKED ISSUE #2002: Base32 support
**Describe the solution you'd like**
Add `base32` support programically.
Base32 is used by two-factor-authentication (ex. TOTP) to generating secret tokens.

Shared functions:
- `string base32Encode( string value )`
- `string base32Decode( string value )`,
- `string encodeString( string "base32", string value )`
- `string decodeString( string "base32", string value )`

### PR DESCRIPTION
**Pull request by fresholia**
Fixes #2002
Adding encodeString base32&base64

Example:
```lua
local str = encodeString("base32", "anan", {})
print(decodeString("base32", str, {})) -- output: anan

local str = encodeString("base64", "anan", {})
print(decodeString("base64", str, {})) -- output: anan

```

When the PR is approved, I will update it on the wiki.

### CODE REVIEW

File: Shared/sdk/SharedUtil.Hash.h [hunk 1]
```diff
@@ -30,7 +30,10 @@ enum class PasswordHashFunction
 
 enum class StringEncryptFunction
 {
-    TEA
+    TEA,
+    BASE32,
+    BASE64,
+    HMAC
```
Comment (patrikjuvonen) [inline Shared/sdk/SharedUtil.Hash.h]: HMAC should be removed from this PR.
```suggestion
```

**Comment by fresholia** [comment]
I guess it could be done as follows, idk
```lua
encodeString("base", string, {radix = 32})
```

**Review by patrikjuvonen — CHANGES_REQUESTED** [review]
Thanks for the pull request! Seems good, I haven't tested this yet but will do so soon.

**Comment by patrikjuvonen** [comment]
> I guess it could be done as follows, idk
> 
> ```lua
> encodeString("base", string, {radix = 32})
> ```

I think base32 and base64 is better for now.

**Review by patrikjuvonen — DISMISSED** [review]

File: Shared/mods/deathmatch/logic/luadefs/CLuaCryptDefs.cpp [hunk 2]
```diff
@@ -318,6 +318,64 @@ int CLuaCryptDefs::EncodeString(lua_State* luaVM)
...
+            case StringEncryptFunction::BASE32:
+            {
+                // Async
+                if (VERIFY_FUNCTION(luaFunctionRef))
+                {
+                    CLuaShared::GetAsyncTaskScheduler()->PushTask<SString>(
+                        [data] {
+                            // Execute time-consuming task
+                            SString result = SharedUtil::Base32encode(data);
+                            return result;
```
Comment (sbx320) [inline Shared/mods/deathmatch/logic/luadefs/CLuaCryptDefs.cpp]: Just `return SharedUtil::Base32encode(data);`

File: Shared/sdk/SharedUtil.Hash.h [hunk 3]
```diff
@@ -30,7 +30,9 @@ enum class PasswordHashFunction
 
 enum class StringEncryptFunction
 {
-    TEA
+    TEA,
+    BASE32,
+    BASE64
 };
```
Comment (sbx320) [inline Shared/sdk/SharedUtil.Hash.h:40]: The `StringEncryptFunction` name is a bit odd due to `base64` not being an encryption method. Maybe we should change this to `StringEncodeFunction`?

**Review by sbx320 — CHANGES_REQUESTED** [review]
Looks good, just some minor changes.

--- 

In general we might want to look into making `Encode/DecodeString` a bit less repetitive. But don't bother with it for this PR.

File: Shared/sdk/SharedUtil.Hash.h (hunk 3 above)
Comment (fresholia) [reply Shared/sdk/SharedUtil.Hash.h:40]: Thanks for review. I'll edit.

**Comment by TheNormalnij** [comment]
base32 should have `alphabet` option i think.
https://github.com/TheNormalnij/mtasa-blue/tree/TheNormalnij/base32

File: Shared/mods/deathmatch/logic/luadefs/CLuaCryptDefs.cpp [hunk 4]
```diff
@@ -318,6 +318,18 @@ int CLuaCryptDefs::EncodeString(lua_State* luaVM)
                 }
                 return 1;
             }
+            case StringEncryptFunction::BASE32:
+            {
+                SString result = SharedUtil::Base32encode(data);
+                lua_pushlstring(luaVM, result, result.length());
```
Comment (Pirulax) [inline Shared/mods/deathmatch/logic/luadefs/CLuaCryptDefs.cpp]: lua::Push(luaVM, SharedUtil::Base32encode(data));

File: Shared/mods/deathmatch/logic/luadefs/CLuaCryptDefs.cpp [hunk 5]
```diff
@@ -318,6 +318,18 @@ int CLuaCryptDefs::EncodeString(lua_State* luaVM)
...
+                return 1;
+            }
+            case StringEncryptFunction::BASE64:
+            {
+                SString result = SharedUtil::Base64encode(data);
+                lua_pushlstring(luaVM, result, result.length());
```
Comment (Pirulax) [inline Shared/mods/deathmatch/logic/luadefs/CLuaCryptDefs.cpp]: Same `lua::Push`

File: Shared/mods/deathmatch/logic/luadefs/CLuaCryptDefs.cpp [hunk 6]
```diff
@@ -397,6 +409,64 @@ int CLuaCryptDefs::DecodeString(lua_State* luaVM)
                 }
                 return 1;
             }
+            case StringEncryptFunction::BASE32:
+            {
+                // Async
+                if (VERIFY_FUNCTION(luaFunctionRef))
+                {
+                    CLuaShared::GetAsyncTaskScheduler()->PushTask<SString>(
+                        [data] {
```
Comment (Pirulax) [inline Shared/mods/deathmatch/logic/luadefs/CLuaCryptDefs.cpp]: `[data = std::move(data)]` in the lambda capture.

**Review by Pirulax — APPROVED** [review]
lgtm.

File: Shared/sdk/SharedUtil.Hash.h (hunk 3 above)
Comment (patrikjuvonen) [reply Shared/sdk/SharedUtil.Hash.h:40]: @cleopatradev This change is still pending.

File: Shared/mods/deathmatch/logic/luadefs/CLuaCryptDefs.cpp [hunk 7]
```diff
@@ -397,6 +407,16 @@ int CLuaCryptDefs::DecodeString(lua_State* luaVM)
...
+                return 1;
+            }
+            case StringEncryptFunction::BASE64:
+            {
+                lua::Push(luaVM, SharedUtil::Base64decode(data));
+                return 1;
+            }
             default:
             {
                 m_pScriptDebugging->LogCustom(luaVM, "Unknown encryption algorithm");
```
Comment (patrikjuvonen) [inline Shared/mods/deathmatch/logic/luadefs/CLuaCryptDefs.cpp]: Let's change this to "Unknown algorithm" instead.
```suggestion
                m_pScriptDebugging->LogCustom(luaVM, "Unknown algorithm");
```

**Review by patrikjuvonen — DISMISSED** [review]

File: Shared/sdk/SharedUtil.Hash.h (hunk 3 above)
Comment (patrikjuvonen) [reply Shared/sdk/SharedUtil.Hash.h:40]: @cleopatradev This is still an issue.

**Comment by Dutchman101** [comment]
Exactly what is needed for this to switch from draft to ready for more reviews or merging? The PR was changed to draft status by @patrikjuvonen, maybe clarify the reason or we can mark it as ready for review again.

**Comment by patrikjuvonen** [comment]
> Exactly what is needed for this to switch from draft to ready for more reviews or merging? The PR was changed to draft status by @patrikjuvonen, maybe clarify the reason or we can mark it as ready for review again.

sbx's request for change is still pending at https://github.com/multitheftauto/mtasa-blue/pull/2016#discussion_r560499522

**Comment by fresholia** [comment]
Wrong branches, wait the new PR

**Pull request closed without merging**